ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -114,6 +115,18 @@ lastK (SecurityParam k) f =
     . fmap f
     . currentChain
 
+-- | The block number of the most recent \"immutable\" block, i.e. the oldest
+-- block we can roll back to. We cannot roll back the block itself.
+--
+-- In the real implementation this will correspond to the block number of the
+-- block at the tip of the Immutable DB.
+immutableBlockNo :: HasHeader blk
+                 => SecurityParam -> Model blk -> Block.BlockNo
+immutableBlockNo (SecurityParam k) =
+        Chain.headBlockNo
+      . Chain.drop (fromIntegral k)
```

Comment:
I think that's right, no off-by-one errors as far as I can see, and clear doc comment to explain.